src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,200 @@
+    end
+
+    # Pseudo-atomic wavefunctions
+    # Currently not used; can be used for initializing the starting wavefunction and as
+    # projectors for projected densities of states, projected wavefunctions, and DFT+U(+V).
+    pswfcs = map(0:lmax - 1) do l
+        pswfcs_l = filter(pseudo["atomic_wave_functions"]) do pswfc
+            pswfc["angular_momentum"] == l
+        end
+        map(pswfc -> pswfc["radial_function"], pswfcs_l)
+    end
+    pswfc_occs = map(0:lmax - 1) do l
+        pswfcs_l = filter(pseudo["atomic_wave_functions"]) do pswfc
+            pswfc["angular_momentum"] == l
+        end
+        map(pswfc -> pswfc["occupation"], pswfcs_l)
+    end
+
+    # Currently not used; can be used for initializing the starting guess density.
+    rhoatom = pseudo["total_charge_density"]
+
+    # Useful precomputed quantities (q-independent parts of Fourier transform integrands)
+    rvlocpzdr = (rgrid .* vloc .+ Zion) .* drgrid
+    r2projsdr = map(projs) do projs_l
+        map(projs_l) do proj
+            ir_cut = lastindex(proj)
+            rgrid[1:ir_cut] .* proj .* drgrid[1:ir_cut]
+        end
+    end
+
+    PspUpf(Zion, lmax, rgrid, drgrid, vloc, vloc_interp, projs, projs_interp, h, pswfcs,
+           pswfc_occs, rhoatom, rvlocpzdr, r2projsdr; identifier)
+end
+
+charge_ionic(psp::PspUpf) = psp.Zion
+
+"""
+    eval_psp_projector_real(psp::PspUpf, i::Number, l::Number, r::Number)
+
+Evaluate the ith Kleinman-Bylander β projector with angular momentum l at real-space
+distance r via linear interpolation on the real-space mesh.
+
+Note: UPFs store `r[i] p_{il}(r[i])`, so r=0 is undefined and will error.
+"""
+function eval_psp_projector_real(psp::PspUpf, i, l, r::T) where {T <: Real}
+    psp.projs_interp[l+1][i](r) / r
+end
+
+"""
+    eval_psp_projector_fourier(psp::PspUPF, i::Number, l::Number, q::Number)
+
+
+Evaluate the radial part of the i-th projector for angular momentum l at the
+reciprocal vector with modulus q:
+p(q) = ∫R^3 p{il}(r) e^{-iqr} dr
+     = 4π ∫{R+} r^2 p_{il}(r) j_l(q r) dr
+     = 4π Σ{i} r[i]^2 p_{il}(r[i]) j_l(q r[i]) dr[i]
+
+Note: UPFs store `r[i] p_{il}(r[i])`
+"""
+function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T)::T where {T <: Real}
+    s = zero(T)
+    @inbounds for ir = eachindex(psp.r2projsdr[l+1][i])
+        s += sphericalbesselj_fast(l, q * psp.rgrid[ir]) * psp.r2projsdr[l+1][i][ir]
+    end
+    4T(π) * s
+end
+
+"""
+     eval_psp_local_real(psp::PspUpf, r::Number)
+
+Evaluate the local potential at real-space distance `r` via linear
+interpolation on the real-space mesh.
+ """
+eval_psp_local_real(psp::PspUpf, r::T) where {T <: Real} = psp.vloc_interp(r)
+
+"""
+    eval_psp_local_fourier(psp::PspUpf, q::Number)
+
+Evaluate the local part of the pseudopotential in reciprocal space
+using a Coulomb correction term -Z/r:
+V(q) = ∫{R^3} (Vloc(r) + Z/r) e^{-iqr} dr
+     = 4π ∫{R+} (Vloc(r) + Z/r) sin(qr)/qr r^2 dr
+     = 4π/q Σ{i} sin(q r[i]) (r[i] V(r[i]) + Z) dr[i]
```

Comment:
this is missing the correction term

Reply:
Clarify here the correction scheme and why it's needed (vloc is long-range)

Reply:
(either here or at the top of the file as a general comment on the algorithms used in the file)

Reply:
In particular it'd be nicer to record the integral used, in case it's useful later